quickwit/quickwit-search/src/leaf.rs

```diff
@@ -879,6 +877,130 @@ impl CanSplitDoBetter {
         }
     }
 
+    /// This function tries to detect upfront which splits contain the top n hits and convert other
+    /// split searches to count only searches. It also optimizes split order.
+    ///
+    /// Returns the search_requests with their split.
+    fn optimize(
+        &self,
+        request: Arc<SearchRequest>,
+        mut splits: Vec<SplitIdAndFooterOffsets>,
+    ) -> Result<Vec<(SplitIdAndFooterOffsets, SearchRequest)>, SearchError> {
+        self.optimize_split_order(&mut splits);
+        // TODO: we maybe want here some deduplication + Cow logic
+        let mut split_with_req = splits
+            .into_iter()
+            .map(|split| (split, (*request).clone()))
+            .collect::<Vec<_>>();
+
+        if request.aggregation_request.is_some() || request.search_after.is_some() {
+            return Ok(split_with_req);
+        }
+        let query_ast: QueryAst = serde_json::from_str(request.query_ast.as_str())
+            .map_err(|err| SearchError::InvalidQuery(err.to_string()))?;
+
+        // TODO: Update the logic to handle start_timestamp end_timestamp ranges
+        if request.start_timestamp.is_some() || request.end_timestamp.is_some() {
+            return Ok(split_with_req);
+        }
+
+        if !matches!(query_ast, QueryAst::MatchAll) {
+            return Ok(split_with_req);
+        }
+
+        // Count the number of splits which contain enough documents
+        let count_required_splits =
```

Comment:
I think we can afford computing it all of the time here and not introduce a closure.

also (but that's a nitpick)... 

I would prefer to have either an iterator style with `scan`:

```rust
        // Count the number of splits which contain enough documents
        let min_required_splits =
            splits.iter()
                .map(|split| split.num_docs as u64)
                // computing the partial sum
                .scan(0u64, |partial_sum: &mut u64, val| {
                    *partial_sum += val;
                    Some(*partial_sum)
                })
                .take_while(|cumulated_count| *cumulated_count < num_requested_docs)
                .count() + 1;
```

or a loop style

```rust
     // Count the number of splits which contain enough documents
        let min_required_splits = {
            let mut cumsum = 0;
            let mut count = 0;
            for split in splits {
                count += 1;
                cumsum += split.num_docs as u64;
                if cumsum >= num_requested_docs {
                    break;
                }
            }
            count
        };
```

The iterator style that mutates stuff captures by the closure is a bit tricky.
That part is a nitpick so feel free to ignore.

Reply:
yes that's better